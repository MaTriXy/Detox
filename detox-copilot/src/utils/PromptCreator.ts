type RankedMethod = {
    method: TestingFrameworkAPIMethod;
    relevance: number;
};

export class PromptCreator {
    constructor(private availableAPI: TestingFrameworkAPI) {
        this.validateAPI();
    }

    private validateAPI() {
        const { actions, assertions, matchers } = this.availableAPI;
        if (!actions || !assertions || !matchers) {
            throw new Error('Invalid TestingFrameworkAPI provided');
        }
    }

    private createBasePrompt(intent: string, viewHierarchy: string): string {
        return `
# Test Code Generation

Generate the JS code to "${intent}" based on the current UI state and the available API methods.

## Context

### View Hierarchy
\`\`\`
${viewHierarchy}
\`\`\`

### Image Snapshot
An image snapshot of the current UI state is attached to this prompt.

### Intent
${intent}

## Guidelines
- Use the most appropriate method(s) from the available API.
- Ensure the code is directly executable in JavaScript.
- If the operation is impossible, explain why in a brief error message (e.g. "A login button element was not found, looks like the user is already logged in").
- Consider the attached image snapshot when determining best course of action.

Return only the executable code or the error message, without any additional formatting or explanation.
        `.trim();
    }

    private rankMethods(methods: TestingFrameworkAPIMethod[], intent: string): RankedMethod[] {
        return methods
            .map(method => ({
                method,
                relevance: this.calculateSemanticSimilarity(method, intent)
            }))
            .sort((a, b) => b.relevance - a.relevance);
    }

    private calculateSemanticSimilarity(method: TestingFrameworkAPIMethod, intent: string): number {
        const methodText = `${method.signature} ${method.description} ${method.example}`;
        const commonWords = this.getCommonWords(methodText, intent);
        const methodVector = this.createVector(methodText, commonWords);
        const intentVector = this.createVector(intent, commonWords);
        return this.cosineSimilarity(methodVector, intentVector);
    }

    private getCommonWords(text1: string, text2: string): Set<string> {
        const words1 = new Set(text1.toLowerCase().split(/\s+/));
        const words2 = new Set(text2.toLowerCase().split(/\s+/));
        return new Set([...words1].filter(word => words2.has(word)));
    }

    private createVector(text: string, vocabulary: Set<string>): number[] {
        const words = text.toLowerCase().split(/\s+/);
        return Array.from(vocabulary).map(word => words.filter(w => w === word).length);
    }

    private cosineSimilarity(vec1: number[], vec2: number[]): number {
        const dotProduct = vec1.reduce((sum, val, i) => sum + val * vec2[i], 0);
        const mag1 = Math.sqrt(vec1.reduce((sum, val) => sum + val * val, 0));
        const mag2 = Math.sqrt(vec2.reduce((sum, val) => sum + val * val, 0));
        return dotProduct / (mag1 * mag2) || 0;  // Return 0 if result is NaN
    }

    private createMethodPrompt(rankedMethod: RankedMethod): string {
        const { method, relevance } = rankedMethod;
        const guidelines = method.guidelines.length > 0
            ? `\n#### Guidelines\n${method.guidelines.map(g => `- ${g}`).join('\n')}`
            : '';

        return `
### ${method.signature}

${method.description}

#### Example
\`\`\`
${method.example}
\`\`\`
${guidelines}
        `.trim();
    }

    private createMatcherPrompt(): string {
        const matcherList = this.availableAPI.matchers
            .map(matcher => `- \`${matcher.signature}\`: ${matcher.description}`)
            .join('\n');

        return `
## Available Matchers

${matcherList}
        `.trim();
    }

    private createPrompt(intent: string, viewHierarchy: string, methodType: 'actions' | 'assertions'): string {
        const basePrompt = this.createBasePrompt(intent, viewHierarchy);
        const rankedMethods = this.rankMethods(this.availableAPI[methodType], intent);
        const suggestedMethods = rankedMethods.slice(0, 3);
        const methodPrompts = rankedMethods.map((rankedMethod, index) =>
            this.createMethodPrompt(rankedMethod, index < 3)
        ).join('\n\n');
        const matcherPrompt = this.createMatcherPrompt();

        return `
${basePrompt}

## Available Methods
${methodPrompts}

## Related Methods

3 most relevant methods, based on semantic similarity:
${suggestedMethods.map(m => `- \`${m.method.signature}\` (relevance: ${m.relevance.toFixed(2)})`).join('\n')}

${matcherPrompt}
        `.trim();
    }

    createActPrompt(action: string, viewHierarchy: string): string {
        return this.createPrompt(action, viewHierarchy, 'actions');
    }

    createExpectPrompt(assertion: string, viewHierarchy: string): string {
        return this.createPrompt(assertion, viewHierarchy, 'assertions');
    }
}
