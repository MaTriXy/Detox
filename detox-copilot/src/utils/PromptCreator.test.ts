import { PromptCreator } from './PromptCreator';

const mockAPI: TestingFrameworkAPI = {
    actions: [
        {
            signature: 'tap(element: Element)',
            description: 'Taps on the specified element.',
            example: 'await element(by.id("button")).tap();',
            guidelines: ['Ensure the element is tappable before using this method.']
        },
        {
            signature: 'typeText(element: Element, text: string)',
            description: 'Types the specified text into the element.',
            example: 'await element(by.id("input")).typeText("Hello, World!");',
            guidelines: ['Use this method only on text input elements.']
        }
    ],
    assertions: [
        {
            signature: 'toBeVisible()',
            description: 'Asserts that the element is visible on the screen.',
            example: 'await expect(element(by.id("title"))).toBeVisible();',
            guidelines: ['Consider scroll position when using this assertion.']
        }
    ],
    matchers: [
        {
            signature: 'by.id(id: string)',
            description: 'Matches elements by their ID attribute.',
            example: 'element(by.id("uniqueId"))',
            guidelines: ['Use unique IDs for elements to avoid conflicts, combine with atIndex() if necessary.']
        }
    ]
};

describe('PromptCreator', () => {
    let promptCreator: PromptCreator;

    beforeEach(() => {
        promptCreator = new PromptCreator(mockAPI);
    });

    it('creates an action prompt correctly', () => {
        const action = 'Tap the submit button';
        const viewHierarchy = '<View><Button testID="submit" title="Submit" /></View>';
        const prompt = promptCreator.createActPrompt(action, viewHierarchy);
        expect(prompt).toMatchSnapshot();
    });

    it('creates an assertion prompt correctly', () => {
        const assertion = 'Check if the success message is visible';
        const viewHierarchy = '<View><Text testID="successMessage">Success!</Text></View>';
        const prompt = promptCreator.createExpectPrompt(assertion, viewHierarchy);
        expect(prompt).toMatchSnapshot();
    });

    it('handles long intents', () => {
        const action = 'Tap the submit button, then wait for the loading indicator to disappear, and finally verify that the success message is displayed';
        const viewHierarchy = '<View><Button testID="submit" /><ActivityIndicator testID="loader" /><Text testID="successMessage" /></View>';
        const prompt = promptCreator.createActPrompt(action, viewHierarchy);
        expect(prompt).toMatchSnapshot();
    });

    it('throws an error for invalid API', () => {
        expect(() => new PromptCreator({} as TestingFrameworkAPI)).toThrow('Invalid TestingFrameworkAPI provided');
    });
});
