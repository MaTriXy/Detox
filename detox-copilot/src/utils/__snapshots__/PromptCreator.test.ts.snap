// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PromptCreator creates an action prompt correctly 1`] = `
"# Test Code Generation

## Context

### View Hierarchy
\`\`\`
<View><Button testID="submit" title="Submit" /></View>
\`\`\`

### Image Snapshot
An image snapshot of the current UI state is attached to this prompt.

### Intent
Tap the submit button

## Guidelines
- Use the most appropriate method(s) from the available API.
- Ensure the code is directly executable in JavaScript.
- If the operation is impossible, explain why in a brief error message (e.g. "A login button element was not found, looks like the user is already logged in").
- Consider the attached image snapshot when determining best course of action.

Return only the executable code or the error message, without any additional formatting or explanation.

## Available Methods
### tap(element: Element) (Suggested, relevance: 1.00)

Taps on the specified element.

#### Example
\`\`\`
await element(by.id("button")).tap();
\`\`\`

#### Guidelines
- Ensure the element is tappable before using this method.

### typeText(element: Element, text: string) (Suggested, relevance: 1.00)

Types the specified text into the element.

#### Example
\`\`\`
await element(by.id("input")).typeText("Hello, World!");
\`\`\`

#### Guidelines
- Use this method only on text input elements.

## Related Methods

3 most relevant methods, based on semantic similarity:
- \`tap(element: Element)\` (relevance: 1.00)
- \`typeText(element: Element, text: string)\` (relevance: 1.00)

## Available Matchers

- \`by.id(id: string)\`: Matches elements by their ID attribute."
`;

exports[`PromptCreator creates an assertion prompt correctly 1`] = `
"# Test Code Generation

## Context

### View Hierarchy
\`\`\`
<View><Text testID="successMessage">Success!</Text></View>
\`\`\`

### Image Snapshot
An image snapshot of the current UI state is attached to this prompt.

### Intent
Check if the success message is visible

## Guidelines
- Use the most appropriate method(s) from the available API.
- Ensure the code is directly executable in JavaScript.
- If the operation is impossible, explain why in a brief error message (e.g. "A login button element was not found, looks like the user is already logged in").
- Consider the attached image snapshot when determining best course of action.

Return only the executable code or the error message, without any additional formatting or explanation.

## Available Methods
### toBeVisible() (Suggested, relevance: 0.94)

Asserts that the element is visible on the screen.

#### Example
\`\`\`
await expect(element(by.id("title"))).toBeVisible();
\`\`\`

#### Guidelines
- Consider scroll position when using this assertion.

## Related Methods

3 most relevant methods, based on semantic similarity:
- \`toBeVisible()\` (relevance: 0.94)

## Available Matchers

- \`by.id(id: string)\`: Matches elements by their ID attribute."
`;

exports[`PromptCreator handles empty view hierarchy 1`] = `
"# Test Code Generation

## Context

### View Hierarchy
\`\`\`

\`\`\`

### Image Snapshot
An image snapshot of the current UI state is attached to this prompt.

### Intent
Tap the submit button

## Guidelines
- Use the most appropriate method(s) from the available API.
- Ensure the code is directly executable in JavaScript.
- If the operation is impossible, explain why in a brief error message (e.g. "A login button element was not found, looks like the user is already logged in").
- Consider the attached image snapshot when determining best course of action.

Return only the executable code or the error message, without any additional formatting or explanation.

## Available Methods
### tap(element: Element) (Suggested, relevance: 1.00)

Taps on the specified element.

#### Example
\`\`\`
await element(by.id("button")).tap();
\`\`\`

#### Guidelines
- Ensure the element is tappable before using this method.

### typeText(element: Element, text: string) (Suggested, relevance: 1.00)

Types the specified text into the element.

#### Example
\`\`\`
await element(by.id("input")).typeText("Hello, World!");
\`\`\`

#### Guidelines
- Use this method only on text input elements.

## Related Methods

3 most relevant methods, based on semantic similarity:
- \`tap(element: Element)\` (relevance: 1.00)
- \`typeText(element: Element, text: string)\` (relevance: 1.00)

## Available Matchers

- \`by.id(id: string)\`: Matches elements by their ID attribute."
`;

exports[`PromptCreator handles long intents 1`] = `
"# Test Code Generation

## Context

### View Hierarchy
\`\`\`
<View><Button testID="submit" /><ActivityIndicator testID="loader" /><Text testID="successMessage" /></View>
\`\`\`

### Image Snapshot
An image snapshot of the current UI state is attached to this prompt.

### Intent
Tap the submit button, then wait for the loading indicator to disappear, and finally verify that the success message is displayed

## Guidelines
- Use the most appropriate method(s) from the available API.
- Ensure the code is directly executable in JavaScript.
- If the operation is impossible, explain why in a brief error message (e.g. "A login button element was not found, looks like the user is already logged in").
- Consider the attached image snapshot when determining best course of action.

Return only the executable code or the error message, without any additional formatting or explanation.

## Available Methods
### tap(element: Element) (Suggested, relevance: 1.00)

Taps on the specified element.

#### Example
\`\`\`
await element(by.id("button")).tap();
\`\`\`

#### Guidelines
- Ensure the element is tappable before using this method.

### typeText(element: Element, text: string) (Suggested, relevance: 1.00)

Types the specified text into the element.

#### Example
\`\`\`
await element(by.id("input")).typeText("Hello, World!");
\`\`\`

#### Guidelines
- Use this method only on text input elements.

## Related Methods

3 most relevant methods, based on semantic similarity:
- \`tap(element: Element)\` (relevance: 1.00)
- \`typeText(element: Element, text: string)\` (relevance: 1.00)

## Available Matchers

- \`by.id(id: string)\`: Matches elements by their ID attribute."
`;
